File: src/main/java/com/zunza/buythedip/cryptocurrency/service/dataInit.java
package com.zunza.buythedip.cryptocurrency.service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Service;

import com.zunza.buythedip.cryptocurrency.client.BinanceClient;
import com.zunza.buythedip.cryptocurrency.client.CoinMarketCapClient;
import com.zunza.buythedip.cryptocurrency.dto.binance.ExchangeInfoDto;
import com.zunza.buythedip.cryptocurrency.dto.binance.SymbolDto;
import com.zunza.buythedip.cryptocurrency.dto.cmc.CryptoCurrencyMetadataDto;
import com.zunza.buythedip.cryptocurrency.dto.cmc.MetadataDetailDto;
import com.zunza.buythedip.cryptocurrency.dto.cmc.UrlsDto;
import com.zunza.buythedip.cryptocurrency.entity.Cryptocurrency;
import com.zunza.buythedip.cryptocurrency.entity.CryptocurrencyMetadata;
import com.zunza.buythedip.cryptocurrency.repository.CryptocurrencyRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
@RequiredArgsConstructor
public class dataInit {

	private final BinanceClient binanceClient;
	private final CoinMarketCapClient cmcClient;
	private final CryptocurrencyRepository cryptoCurrencyRepository;

	// @PostConstruct
	private void cryptoCurrencyDataInit() {
		ExchangeInfoDto exchangeInfo = binanceClient.getExchangeInfo();
		List<SymbolDto> filteredSymbols = filterUsdtPairsAndTrading(exchangeInfo.getSymbols());

		log.info(String.valueOf(filteredSymbols.size()));
		filteredSymbols.forEach(symbol ->
			log.info("{} / {}", symbol.getBaseAsset(), symbol.getStatus())
			);

		List<String> baseAssetList = filteredSymbols.stream()
			.map(SymbolDto::getBaseAsset)
			.toList();

		List<List<String>> partitionList = partitionList(baseAssetList, 100);
		List<CryptoCurrencyMetadataDto> result = new ArrayList<>();

		for (List<String> list : partitionList) {
			String symbolParam = String.join(",", list);
			CryptoCurrencyMetadataDto metadata = cmcClient.getCryptoCurrencyMetadata(symbolParam);

			result.add(metadata);
		}

		List<Cryptocurrency> cryptocurrencyList = new ArrayList<>();

		for (CryptoCurrencyMetadataDto metadata : result) {
			for (String s : metadata.getData().keySet()) {

				List<MetadataDetailDto> metadataDetailDtos = metadata.getData().get(s);
				if (metadataDetailDtos == null || metadataDetailDtos.isEmpty()) continue;

				MetadataDetailDto detail = metadataDetailDtos.get(0);
				if (detail == null) continue;

				UrlsDto urls = detail.getUrls();
				CryptocurrencyMetadata cryptoCurrencyMetadata = CryptocurrencyMetadata.of(
					detail.getLogo(), detail.getDescription(), urls.getWebsite(),
					urls.getTwitter(), urls.getExplorer(), detail.getTagNames());

				Cryptocurrency cryptoCurrency = Cryptocurrency.of(detail.getName(), detail.getSymbol(), "TRADING", cryptoCurrencyMetadata);
				cryptocurrencyList.add(cryptoCurrency);
			}
		}

		cryptoCurrencyRepository.saveAll(cryptocurrencyList);
		log.info("==============COMPLETE===================");
	}

	private List<SymbolDto> filterUsdtPairsAndTrading(List<SymbolDto> symbols) {
		return symbols.stream()
			.filter(symbol -> symbol.getQuoteAsset().equals("USDT"))
			.filter(symbol -> symbol.getStatus().equals("TRADING"))
			.toList();
	}

	private List<List<String>> partitionList(List<String> list, int size) {
		List<List<String>> result = new ArrayList<>();
		for (int i = 0; i < list.size(); i += size) {
			result.add(list.subList(i, Math.min(i + size, list.size())));
		}
		return result;
	}
}

File: src/main/kotlin/com/zunza/buythedip/external/DataInit.kt
package com.zunza.buythedip.external

import com.zunza.buythedip.crypto.entity.Crypto
import com.zunza.buythedip.crypto.entity.CryptoMetadata
import com.zunza.buythedip.crypto.repository.CryptoMetadataRepository
import com.zunza.buythedip.crypto.repository.CryptoRepository
import com.zunza.buythedip.external.binance.client.BinanceClient
import com.zunza.buythedip.external.binance.dto.Symbol
import com.zunza.buythedip.external.coinmarketcap.client.CoinMarketCapClient
import com.zunza.buythedip.external.coinmarketcap.dto.Metadata
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.annotation.PostConstruct
import org.springframework.stereotype.Component

private val logger = KotlinLogging.logger {  }

//@Component
class DataInit(
    private val binanceClient: BinanceClient,
    private val coinMarketCapClient: CoinMarketCapClient,
    private val cryptoRepository: CryptoRepository,
    private val cryptoMetadataRepository: CryptoMetadataRepository
) {
//    @PostConstruct
    fun dataInit() {
        logger.info { "==============START===============" }

        val exchangeInfo = binanceClient.getExchangeInfo()
        val filtered = filter(exchangeInfo.symbols)

        val baseAssets = filtered.map { it.baseAsset }
        val partitionList = partitionList(baseAssets, 100)

        val result = mutableListOf<Metadata>()
        for (list in partitionList) {
            val symbols = list.joinToString(",")
            logger.info { symbols }
            val metadata = coinMarketCapClient.getInfo(symbols)
            result.add(metadata)
        }

        for (metadata in result) {
            for (s in metadata.data.keys) {

                val details = metadata.data[s]
                if(details.isNullOrEmpty()) continue
                val detail = details[0]
                val urls = detail.urls

                val crypto = Crypto(
                    name = detail.name,
                    symbol = detail.symbol,
                    logo = detail.logo,
                )
                val savedCrypto = cryptoRepository.save(crypto)
                logger.info { savedCrypto }

                val cryptoMetadata = CryptoMetadata(
                    crypto = savedCrypto,
                    description = detail.description ?: "",
                    website = urls.website,
                    twitter = urls.twitter,
                    explorer = urls.explorer,
                    tags = detail.tags ?: emptyList()
                    )
                val savedMeta = cryptoMetadataRepository.save(cryptoMetadata)
                logger.info { savedMeta }
            }
        }
        logger.info { "==============END===============" }
    }

    private fun filter(symbols: List<Symbol>): List<Symbol> {
        return symbols
            .filter { it.quoteAsset == "USDT" }
            .filter { it.status == "TRADING" }
    }

    private fun partitionList(list: List<String>, size: Int): MutableList<List<String>> {
        val result = mutableListOf<List<String>>()
        for (i in 0 until list.size step size) {
            result.add(list.subList(i, (i + size).coerceAtMost(list.size)))
        }
        return result
    }
}
